File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        res = total = 0
        sum_hash = {}
        
        for num in nums:
            total += num
            
            if total == goal:
                res += 1
            
            if (total - goal) in sum_hash:
                res += sum_hash[total - goal]
            
            if total in sum_hash:
                sum_hash[total] += 1
            else:
                sum_hash[total] = 1

        return res

File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.rb
# @param {Integer[]} nums
# @param {Integer} goal
# @return {Integer}
def num_subarrays_with_sum(nums, goal)
    res = 0
    sum_hash = Hash.new(0)
    total = 0

    nums.each do |num|
        total += num
        
        res += 1 if total == goal
        
        res += sum_hash[(total - goal)] 
        sum_hash[total] += 1
    end
    res
end
